### ISSUE TITLE
Stuck at database installation step : CRSF Token mismatch error

### ISSUE THREAD

**Issue by Syam** [InvoiceShelf/InvoiceShelf#213]
**Describe the bug**
On a fresh docker compose install : 
Can't submit and continue the "Site URL & Database" installation step.
Get a "CSRF token mismatch" error + something went wrong error.
Tried both Sqlite and Mysql connections.

**Expected behavior**
Should continue without error.

**Screenshots**
![image](https://github.com/user-attachments/assets/e3515d61-e6fb-4cf4-a728-17758952edf3)
![image](https://github.com/user-attachments/assets/d449253b-3ffe-433a-9829-0442789ad118)

**Please complete the following information:**
- InvoiceShelf version: 2.0.0 (commit ddca5a7)
- PHP version: 8.2.24
- Database type and version: MariadDB 10 and SQLite

**Optional info**
- OS:  Debian 12
- Browser: Chrome version 130.0.6723.117

**Comment by Syam** [comment]
Maybe an issue with cache clearing on the early controller, that reset the CRSF TOKEN cookie ?

https://github.com/InvoiceShelf/InvoiceShelf/blob/ddca5a7b6d871cf7c4d48ae2323a6922ca1f07d8/app/Http/Controllers/V1/Installation/DatabaseConfigurationController.php#L27

**Comment by Syam** [comment]
Another hint is Laravel version 11 breaking the CRSF check :

> In my case it was problem with VerifyCsrfToken (leftover from laravel 10). In 11 version, laravel loads it automatically by Illuminate\Foundation\Http\Middleware\ValidateCsrfToken.
> So, i think if you have in your code VerifyCsrfToken class laravel will load it two times and something going wrong.
> 

Source : https://laracasts.com/discuss/channels/laravel/laravel-shift-10x-to-11x-csrf-token-mismatch?page=1&replyId=932223

**Comment by rihards-simanovics** [comment]
Thanks @loduis this indeed should solve the issue. Unfortunately, CSRF error is extremely common during installs and re-installs of the app as the cookies don't match. @Syam try clearing site data, and this should solve your problem.

Closing as solved.

**Comment by rihards-simanovics** [comment]
> a cookie must be being created with the domain name without a dot and in the other with the dot

Yes, it's likely managed by the core of the app which is written in php & Laravel and most likely not the static frontend.

I am already looking through the code from the frontend in preparation for migration to React Native, where first phase would be to replace the installation screens with React Native followed by the app itself.

**Comment by loduis** [comment]
@rihards-simanovics I have a question for you, are you planning to migrate from Vue to React?

**Comment by rihards-simanovics** [comment]
> @rihards-simanovics I have a question for you, are you planning to migrate from Vue to React?

Yes, I'll be doing that in the next year or so, RN should simplify the frontend management as we could generate both mobile app and static web from one codebase. The transition will be slow, as a lot of testing will be required, so in the interim feel free to submit PRs for Vue fixes, I think I've seen one of your PRs already.

**Comment by kylelliott** [comment]
Hi All

i seem to have same problem tho even clearing cache still not fixing the problem

**Comment by loduis** [comment]
@kylelliott You need remove the cookies

**Comment by kylelliott** [comment]
doesnt work

**Comment by yasdanibrenes** [comment]
Using Herd the application worked out clearing cache or using another browser with the CRSF TOKEN.


But using XAMPP gives me a 404 page error and installer doesn't just work.
![Screenshot 2024-11-17 101854](https://github.com/user-attachments/assets/3ba24fe8-11cb-437d-9750-3aeef201d4cb)


Tried with my hosting Siteground and same error, 404 page error when entering the route https://sub.domain/InvoiceShelf/public/installation
![Screenshot 2024-11-17 101557](https://github.com/user-attachments/assets/c13faf43-d38d-4c8e-b4db-4827e4646c4c)

**Comment by rihards-simanovics** [comment]
@yasdanibrenes why you are going inside /public/, you should point the Webroot to /public/ dir. This is likely why it's not working because that route doesn't exist in the routing table.

**Comment by yasdanibrenes** [comment]
Great, that makes sense and I was able to fix it with .htaccess adding the following in the root folder

`RewriteEngine On
RedirectMatch ^/$ /InvoiceShelf/public/`

Thanks @rihards-simanovics

**Comment by gdarko** [comment]
hey @kylelliott  - Please open browser console and see what is the error. Thanks.

**Comment by kylelliott** [comment]
> hey @kylelliott  - Please open browser console and see what is the error. Thanks.

![image](https://github.com/user-attachments/assets/cce9e3f7-a499-43fd-9dcf-dc5a246305c4)

**Comment by gdarko** [comment]
> > hey @kylelliott  - Please open browser console and see what is the error. Thanks.
> 
> ![image](https://private-user-images.githubusercontent.com/23058718/386990096-cce9e3f7-a499-43fd-9dcf-dc5a246305c4.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE4NzI1MjcsIm5iZiI6MTczMTg3MjIyNywicGF0aCI6Ii8yMzA1ODcxOC8zODY5OTAwOTYtY2NlOWUzZjctYTQ5OS00M2ZkLTlkY2YtZGM1YTI0NjMwNWM0LmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExN1QxOTM3MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNWEyNmE2NjczYzU1ODgwMmQzMTEyN2NhNTNjMmJmOTgyNDE4MGMwNmE0Y2Q3MmUzMmE4ZTgxM2JmMDkzMzgxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.y-dMYr2AzlWKtXoJ41SvFN8TW72koOm3REvUMBmbWsY)

Thanks. There are a few things you could try:

- Check your log file in storage/logs
- Check the network tab and click on the red colored request in Browser console

Also, how do you deploy your site, do you use the Docker image or manually?

**Comment by kylelliott** [comment]
You using docker and have tried all 3 compose and having the same problem soon as reboot I can on longer login 

![image](https://github.com/user-attachments/assets/f837aa96-e987-4ee0-a7b6-f6524b08148e)

**Comment by rihards-simanovics** [comment]
@kylelliott Github issues might now be the best place for debugging this kind of issue as it might have something to do with your env. and will require quite a bit of back and forth. You are welcome to either open a question in [discussions](https://github.com/InvoiceShelf/InvoiceShelf/discussions/new/choose) or join our discord server and ask your question there. Either way, we will try to help best we can.

**Comment by SvenKribitz** [comment]
![image](https://github.com/user-attachments/assets/4a182ea5-16f2-4197-af37-cc04ec01dfc7)
![image](https://github.com/user-attachments/assets/1d0b6361-4522-478c-a10a-289946d315e0)
About the same here - but I actually got it to run before. I restarted my server and now I can't access InvoiceShelf anymore.

**Comment by rihards-simanovics** [comment]
> About the same here - but I actually got it to run before. I restarted my server and now I can't login to InvoiceShelf anymore.

@gdarko, Huston we have a problem.

I've actually seen this issue before but couldn't reliably replicate it, it seems that it mostly affect platforms that re-create containers instead of restarting them. Coolify is really bad with this. It looks like the entry point script that runs on container start, doesn't correctly pull the env vars and just falls back on the original ones... Which then of course breaks the CSRF as all domains are messed up.

**Comment by loduis** [comment]
@rihards-simanovics This is the problem, comment this line.

```php
// Artisan::call('key:generate --force');
```

![image](https://github.com/user-attachments/assets/d9900d64-f53b-459f-baa2-7e0c7e114136)

https://github.com/InvoiceShelf/InvoiceShelf/pull/224/commits/e253fc773d1ce0c1aac9836c1db89175f221fe63#diff-428ef33e32f0a9a3e786792caabbd56a267f801faa48b2f11d87726f18b84b91R35

**Comment by rihards-simanovics** [comment]
@gdarko, @loduis perhaps this is an extra part of another issue then. In Coolify almost all values get reset to default when I restart the container.

TL;DR is that on restart both `SESSION_DOMAIN` and `SANCTUM_STATEFUL_DOMAINS` among others get reset to localhost (or other default state) despite me providing explicit values by reference to APP_URL, which itself is getting its value from Coollify FQDN env var.

In case it helps, here is a screen recording of the issue, it doesn't give me CSRF, but it does generate 401 (Unauthorised) on bootstrap error: 

https://github.com/user-attachments/assets/43c51698-72f6-4ccd-9e0b-9082cc4a3401

**Comment by SvenKribitz** [comment]
Still can't login.
![image](https://github.com/user-attachments/assets/b5992253-0589-4fce-b4b8-25638353f7f4)
Deleted my cache and tried multiple devices.
Was able to login once and when the server restarted, I'm not able to anymore.

**Comment by mchev** [comment]
> Still can't login.

Did you try deleting the files in `storage/framework/sessions`?

**Comment by SvenKribitz** [comment]
> > Still can't login.
> 
> Did you try deleting the files in `storage/framework/sessions`?

Yes, I deleted the content within that directory. Still the same error.

**Comment by mchev** [comment]
> > > Still can't login.
> > 
> > 
> > Did you try deleting the files in `storage/framework/sessions`?
> 
> Yes, I deleted the content within that directory. Still the same error.

I have two remaining suggestions to try:
- run `php artisan optimize:clear`
- Adjust the `SESSION_DRIVER` in your .env file. You can set it to either 'cookie' or 'database' (if you have the necessary migrations in place). Then re run `php artisan config:clear`

**Comment by SvenKribitz** [comment]
> > > > Still can't login.
> > > 
> > > 
> > > Did you try deleting the files in `storage/framework/sessions`?
> > 
> > 
> > Yes, I deleted the content within that directory. Still the same error.
> 
> I have two remaining suggestions to try:
> 
> * run `php artisan optimize:clear`
> * Adjust the `SESSION_DRIVER` in your .env file. You can set it to either 'cookie' or 'database' (if you have the necessary migrations in place). Then re run `php artisan config:clear`

Sadly none of those suggestions helped. The error remains.

**Comment by mchev** [comment]
Can you share a screenshot of your console and also the cookie list like in https://github.com/InvoiceShelf/InvoiceShelf/issues/213#issuecomment-2477617684 ?

Did you rebuild the docker image ?
What is the SESSION_SRIVER value ?
If the APP_KEY has changed, did you try forgot password ?

Did you try the discord channel for support ?

**Comment by SvenKribitz** [comment]
> Can you share a screenshot of your console and also the cookie list like in [#213 (comment)](https://github.com/InvoiceShelf/InvoiceShelf/issues/213#issuecomment-2477617684) ?
> 
> Did you rebuild the docker image ? What is the SESSION_SRIVER value ? If the APP_KEY has changed, did you try forgot password ?
> 
> Did you try the discord channel for support ?
![image](https://github.com/user-attachments/assets/3288fc95-d01e-40be-9062-1deb6237d85e)
SESSION_DRIVER was on file for SQLite, is on Cookie now.
It just runs locally, I don't use an email server for this one.

I will try the Discord channel, didn't know there was one.

**Comment by rihards-simanovics** [comment]
> I also deploy straight up from https://github.com/invoiceshelf/docker using docker-compose.sqlite.yml

@SvenKribitz there is a known issue where restarting the docker container could result in reset value in `.env` file.

Make sure to exec into the container with an interactive shell `docker exec -it /bin/bash [container id]` (you can get container id by running `docker ps` and looking for container with the name of InvoiceShelf), then inside the container run `apt update && apt install nano` and using  the installed nano, edit the `.env` file to correct the domains: `nano .env`. Once you finished editing the file user `Ctrl + S` to save the file and `Ctrl + X` to close the nano editor and `exit` to exit from the interactive shell.

These are the values you want to check:
```properties
APP_URL=https://invoice.test.io
SESSION_DOMAIN=invoice.test.io
SANCTUM_STATEFUL_DOMAINS=invoice.test.io
```
make sure that the domain name in the values is equal to your domain name and not `localhost` or `localhost:90`

check my video I uploaded in the above comment here for reference on what to do: https://github.com/InvoiceShelf/InvoiceShelf/issues/213#issuecomment-2501063391

**Comment by SvenKribitz** [comment]
> > I also deploy straight up from https://github.com/invoiceshelf/docker using docker-compose.sqlite.yml
> 
> @SvenKribitz there is a known issue where restarting the docker container could result in reset value in `.env` file.
> 
> Make sure to exec into the container with an interactive shell `docker exec -it /bin/bash [container id]` (you can get container id by running `docker ps` and looking for container with the name of InvoiceShelf), then inside the container run `apt update && apt install nano` and using the installed nano, edit the `.env` file to correct the domains: `nano .env`. Once you finished editing the file user `Ctrl + S` to save the file and `Ctrl + X` to close the nano editor and `exit` to exit from the interactive shell.
> 
> These are the values you want to check:
> 
> ```ini
> APP_URL=https://invoice.test.io
> SESSION_DOMAIN=invoice.test.io
> SANCTUM_STATEFUL_DOMAINS=invoice.test.io
> ```
> 
> make sure that the domain name in the values is equal to your domain name and not `localhost` or `localhost:90`
> 
> check my video I uploaded in the above comment here for reference on what to do: [#213 (comment)](https://github.com/InvoiceShelf/InvoiceShelf/issues/213#issuecomment-2501063391)
Awesome, that solves it.
Thought the other issue was rather unrelated because it didn't happen on setup, only restart.
Thank you very much!

**Comment by rihards-simanovics** [comment]
> Thought the other issue was rather unrelated because it didn't happen on setup, only restart

Hey @SvenKribitz, yes that's it. when you set it up the first time the setup wizard does everything correctly but next time you restart it revers the .env values to the original `localhost` and `localhost:90` specifically for `SESSION_DOMAIN` and the `SANCTUM_STATEFUL_DOMAINS`. Since I already cloned a fork of the docker repo. I'll have a look into it some time before the end of year, but no guarantees, as there might be a need for some debugging and I'm amidst the Uni project assessments, so worst case scenario January next year unless someone else fixes it before.

**Comment by christoferd** [comment]
This just gets worse as I have disabled CSRF checking for "installation" and removed from Boostrap 'web', and still get the same error.

**class VerifyCsrfToken**

![Image](https://github.com/user-attachments/assets/8b97ffb5-d231-42f2-94d5-8912ba43ae14)

**bootstrap/app.php**

![Image](https://github.com/user-attachments/assets/4b8f8f88-c43b-42fb-9744-29ca77712bd7)

![Image](https://github.com/user-attachments/assets/90a5e98b-5ff4-4d67-b1b5-47d696eb0775)

**config/sanctum.php**

![Image](https://github.com/user-attachments/assets/781a37b7-c0c0-4f39-8e92-217f2a3ef6c3)

How am I still seeing that message?

**Comment by rihards-simanovics** [comment]
Hey @christoferd, please refer to this issue, which contains all known issues. Even though it's for v2.0.0, row 1 it is still relevant to the latest version: #218.

I am not curtain why your first action would be to remove the code in this case as it only sweeps the problem under the rug. Please examine your environment and look through known issues to find a solution as this problem has been well documented.

**Comment by christoferd** [comment]
Hi @rihards-simanovics thanks for responding

> I am not curtain why your first action would be to remove the code in this case as it only sweeps the problem under the rug.

I just wanted to install to see what it does. Just to try it, experience it, and learn from the code. CSRF on my local machine is not an issue.

PS: I eventually got it running by doing more "sweeping under the rug"

**Comment by rihards-simanovics** [comment]
I am locking this issue as the `CSRF` has been well documented in known bugs, and solutions already exist for both Docker and Traditional hosting. Please refer to #218.

**Comment by rihards-simanovics** [comment]
I would like to reiterate for those who haven't read the comments above, or are still stuck.

If you're experiencing issues with CSRF, check the .env file or Docker Compose YAML for environment variables, as these determine which domains the application trusts. **NOTE**: For Docker, after changing `docker-compose.yaml`, you'll need to restart the application

```properties
APP_URL=http://192.168.1.200
APP_URL=http://192.168.1.200:8080
APP_URL=http://199.199.1.199 # not recommended, run behind reverse proxy with SSL
APP_URL=http://invoiceshelf.acme.com # not recommended, try adding SSL
APP_URL=https://invoiceshelf.acme.com
APP_URL=https://invoiceshelf.acme.com:8080
```
`APP_URL` should have:
1. protocol `http` or `https` if you are accessing it via reverse proxy with SSL, then it will be `https`, if via local IP, then `http`.
2. IP address of the server or the domain name
3. (optional) If you are using a custom port such as `:8080` or another.

```properties
SESSION_DOMAIN=192.168.1.200
SESSION_DOMAIN=192.168.1.200:8080
SESSION_DOMAIN=invoiceshelf.acme.com
SESSION_DOMAIN=invoiceshelf.acme.com:8080
```
`SESSION_DOMAIN` should have:
1. IP address of the server or the domain name
2. (optional) If you are using a custom port such as `:8080` or another.

```properties
# eg 1 - local domains only
SANCTUM_STATEFUL_DOMAINS=localhost,localhost:3000,127.0.0.1,127.0.0.1:8000,::1
# eg 2 - external domain or port
SANCTUM_STATEFUL_DOMAINS=invoiceshelf.acme.com,invoiceshelf.acme.com:8080
# eg 3 - one domain
SANCTUM_STATEFUL_DOMAINS=invoiceshelf.acme.com
```
`SANCTUM_STATEFUL_DOMAINS` should have:
1. IP address of the server or the domain name
1.1. (optional) If you are using a custom port such as `:8080` or another.
2. (optional) If you are planning to access the application from different domains, you must add a comma between each domain
